lib/vector-core/buffers/src/test/model/mod.rs

```diff
@@ -0,0 +1,181 @@
+    inner: Variant,
+}
+
+impl VariantGuard {
+    fn new(variant: Variant) -> Self {
+        match variant {
+            Variant::Memory { .. } => VariantGuard { inner: variant },
+            #[cfg(feature = "disk-buffer")]
+            Variant::Disk {
+                max_size,
+                when_full,
+                name,
+                ..
+            } => {
+                // SAFETY: We allow tempdir to create the directory but by
+                // calling `into_path` we obligate ourselves to delete it. This
+                // is done in the drop implementation for `VariantGuard`.
+                let data_dir = tempdir::TempDir::new_in(std::env::temp_dir(), &name)
+                    .unwrap()
+                    .into_path();
+                VariantGuard {
+                    inner: Variant::Disk {
+                        max_size,
+                        when_full,
+                        data_dir,
+                        name,
+                    },
+                }
+            }
+        }
+    }
+}
+
+impl AsRef<Variant> for VariantGuard {
+    fn as_ref(&self) -> &Variant {
+        &self.inner
+    }
+}
+
+impl Drop for VariantGuard {
+    fn drop(&mut self) {
+        match &self.inner {
+            Variant::Memory { .. } => { /* nothing to clean up */ }
+            #[cfg(feature = "disk-buffer")]
+            Variant::Disk { data_dir, .. } => {
+                // SAFETY: Here we clean up the data_dir of the inner `Variant`,
+                // see note in the constructor for this type.
+                std::fs::remove_dir_all(data_dir).unwrap();
+            }
+        }
+    }
+}
+
+/// This test models a single sender and a single receiver pushing and pulling
+/// from a common buffer. The buffer itself may be either memory or disk. We use
+/// the raw `futures::sink::Sink` and `futures::stream::Stream` interface,
+/// avoiding the need to model the runtime in any way. This is, then, the buffer
+/// as a runtime will see it.
+///
+/// Acks are not modeled yet. I believe doing so would be a straightforward
+/// process.
+#[test]
+fn model_check() {
+    fn inner(variant: Variant, actions: Vec<Action>) -> TestResult {
+        if !check(&variant) {
+            return TestResult::discard();
+        }
+
+        let guard = VariantGuard::new(variant);
+        let mut model: Box<dyn Model> = match guard.as_ref() {
+            Variant::Memory { .. } => Box::new(InMemory::new(guard.as_ref(), 1)),
+            #[cfg(feature = "disk-buffer")]
+            Variant::Disk { .. } => Box::new(OnDisk::new(guard.as_ref())),
+        };
+
+        let rcv_waker = noop_waker();
+        let mut rcv_context = Context::from_waker(&rcv_waker);
+
+        let snd_waker = noop_waker();
+        let mut snd_context = Context::from_waker(&snd_waker);
+
+        let (tx, mut rx, _) = crate::build::<Message>(guard.as_ref().clone()).unwrap();
+
+        let mut tx = tx.get();
+        let sink = tx.as_mut();
```

Comment:
Maybe just pin `sink` here instead of all the spots below?

Reply:
You know, I meant to come back to this. So we can't just move the `Pin` into the loop and it's not clear to me what the `Clone` implementation of `Pin` actually does, whether it clones the interior `P` or not.